### DISCUSSION TITLE
JS error: [https://myinstallation.com/ : 1] Uncaught SyntaxError: Unexpected token '<'

### DISCUSSION THREAD

**Discussion by mperklin** [Novik/ruTorrent#2454] [Q&A] [answered]
I've used ruTorrent for nearly a decade on another server, and have recently built a new one to replace it, but I can't seem to get it to work. I've spent the last 3 days trying to debug why I could not get ruTorrent working, but have been unsuccessful. 

I don't know if what I'm seeing is related to some kind of bug in the most-recent 4.01 version of ruTorrent, or some kind of compatibility issue with the latest version of Ubuntu, or simply my mistakes.

So, I've decided to come here for help.

The title of this post shows the error message I see in ruTorrent's "log" tab when attempting to load ruTorrent:
```JS error: [https://myinstallation.com/ : 1] Uncaught SyntaxError: Unexpected token '<'```

Here is version information that I'm working with:

- Ubuntu 22.10 kinetic
- Apache/2.4.54
- PHP 8.1.7-1ubuntu3.2
- rtorrent 0.9.8
- ruTorrent 4.0.1-hotfix

### rtorrent
Rtorrent is running successfully as a system service using a dedicated user named `rtorrent`.
It is configured with `network.scgi.open_local = /var/run/rtorrent/rtorrent.sock` and a `execute.nothrow` that changes the permissions for this socket file to `770`, allowing members of the `rtorrent` group to access it.

I have confidence this is configured correctly because:

- I am able to use the socket successfully with the Flood user interface to interact with rtorrent. (Flood is running as a system service using a dedicated user named `flood` that is a member of the `rtorrent` group. This group membership allows the `flood` user to access the socket successfully.). **yes, I have stopped flood** to make sure nothing else is accessing the socket except for ruTorrent.
- I am also able to confirm that I do not encounter any `Permission Denied` errors when executing this command: `echo system.listMethods | sudo -u www-data nc -U /var/run/rtorrent/rtorrent.sock` which further confirms the socket is created and accessible.

### Apache
Apache is running successfully as a system service using a dedicated user named `www-data`. It defines a site that serves data from the `/var/www/html` directory which contains the ruTorrent package.

I have confirmed that PHP is working successfully when serving this directory by creating a `test.php` file that contains `<?php phpinfo(); ?>`. PHP info is successfully served, and it includes details of the `mbstring` package that has been installed.

Here is the relevant piece of Apache's site configuration:

```
        <VirtualHost _default_:443>
		ServerAdmin admin@myinstallation.com
                ServerName myinstallation.com:443
		DocumentRoot /var/www/html

                ProxyPass /RPC2 unix:/var/run/rtorrent/rtorrent.sock|scgi://127.0.0.1
                <Directory /var/www/html>
                    AuthType Basic
                    AuthName "myinstallation.com"
                    AuthUserFile /var/www/.authrt
                    Require valid-user
                    Options -Indexes -MultiViews
                    AllowOverride all
                    Require all granted
                </Directory>
        </VirtualHost>
```

I am unable to use an scgi port vs. a scgi socket because the `libapache2-mod-scgi` package is **not** available for Ubuntu kinetic. So I can't use the `SCGIMount` directive. Instead, I'm trying to use a socket via the `ProxyPass` directive per the [instructions referenced here](https://github.com/Novik/ruTorrent/wiki/WebSERVER#mod-proxy-scgi).

### ruTorrent
I've made sure the files untarred at `/var/www/html/` are all owned by `root:www-data` (recursively) so they are accessible by the `www-data` user.

I've made the following conf changes to `/var/www/html/conf/config.php`:

- `$log_file = '/var/log/apache2/rutorrent_errors.log';` - No log file is created at this path
- `$forbidUserSettings = true;` - I did this to remove any confusion related to multiple config.php files for various users, and ensure I'm using the one true config.php
- `$scgi_port = 0;`
- `$scgi_host = "unix:///var/run/rtorrent/rtorrent.sock";`

I've tried outputting the `$log_file` to various locations on disk, and in every case, no error log is created and no debug information is available for me to divine what is going wrong.

I'm making an assumption that ruTorrent simply can't connect to rtorrent via the socket, and it's returning some kind of error message which can't be parsed properly, resulting in the `Uncaught SyntaxError: Unexpected token '<'` error.
I am assuming that if I can get it to connect, all will be well.

I could absolutely be wrong in my assumptions.

### Summary
I'm pulling my hair out trying to figure out why I can't get ruTorrent to successfully use rtorrent's socket. Any questions you can ask or help you can offer will be greatly appreciated. 

I also don't know how to use the `httprpc` plugin since its [documentation](https://github.com/Novik/ruTorrent/wiki/PluginHTTPRPC) provides no direction on its configuration, and the link to the ruTorrent forums within those docs lead to a dead site.

Please help.
Thank you!

**Comment by stickz** [comment, marked as answer]
Clear your web browser cache. Also try upgrading to v4.0.2 released a couple hours ago, for additional stability improvements.
https://github.com/Novik/ruTorrent/releases/tag/v4.0.2

**Reply by mperklin** [reply]
🤦‍♂️ thank you. I didn't realize it was caching the old page despite me refreshing it after each change. Ugh there go a few hours of my life.

Thank you.
